docs/src/writing-constructs.rst

```diff
@@ -0,0 +1,169 @@
+* Validate as early as possible; throw an `Error` in the constructor if the
+  parameters don't make sense. Only if you want to validate mutations that can
+  occur after construction time, override the `validate()` method.
+* Avoid unneeded hierarchy in props; try to keep the props interface flat to
+  help discoverability.
+* Don't expose **Token** s to your consumers; tokens are an implementation
+  mechanism for one of two purpose: representing CloudFormation intrinsic
+  values, or representing lazily evaluated values. They can be used for
+  implementation purposes, but use more specific types as part of your public
+  API.
+* Don't use type unions (**string | BucketName**); they do not translate well
+  across **jsii**. Instead, either just accept a **string** or a **BucketName**.
+  One can always be translated into the other (either by Token stringification
+  or by constructing a **BucketName** from a literal).
+* Accept objects instead of ARNs; when accepting other resources as parameters,
+  declare your property as **queue: QueueRef** instead of **queueArn: Arn**.
+  This makes snapping objects together work logical, and allows you to
+  query/modify the incoming resource as well. The latter is particularly
+  useful if something about IAM permissions need to be set, for example.
+* Use **XxxRef** instead of **Xxx** to allow imports to work; when accepting
+  resource parameters, accept the **XxxRef** variant of the class (if
+  available). This makes it possible for users to pass resources that aren't
+  defined in the same CDK app.
+* Name wrapped resource constructs **Resource** or **Default**; The "main"
+  resource that an AWS Construct represents should use the ID **Resource**,
+  for higher-level wrapping resources you will generally use **Default**
+  (resources named **Default** will inherit their parent's logical ID,
+  while resources named **Resource** will have a distinct logical ID but
+  the human-readable part of it will not show the "Resource" part).
+
+
+Code Organization
+=================
+
+* Your package is named ``aws-xxx`` if it represents the canonical AWS
+  Construct Library for this service; otherwise we recommend starting with
+  ``cdk-``, but you are free to pick a pleasing name.
+* Code goes under **lib/**, tests go under **test/**.
+* Entry point should be **lib/index.ts** and should only contain ``export`` s
+  for other files.
+* No need to put every class in a separate file. Try to think of a
+  reader-friendly organization of your source files.
+* If you want to make package-private utility functions, put them in a file
+  that is *not exported* from **index.ts** and use that file as normal.
+* Free-floating functions cannot be accessed through **jsii** (i.e., from
+  languages other than TypeScript and JavaScript). Don't use them for
+  public features of your construct library.
+* Document all public APIs with doc comments. Document defaults using the **@default**
```

Comment:
Using JSDoc syntax. Maybe we actually want to have JSII documentation on the docs we accept or not (for example, the code highlighting and all is ReST which isn't what most people are used to... I filed awslabs/jsii#197 about that...